### ISSUE TITLE
compilation of files with special charaters in Docker containers

### ISSUE THREAD

**Issue by prasadtalasila** [AutolabJS/AutolabJS#187]
Previously, it has been observed that if submission files contain any special characters, java compilation on docker containers fails. The compilation fails even if same version of Java runs the program successfully natively on top of Ubuntu Linux.

This problem was observed in September-2016. We are not sure if the problem is still there.

**Comment by AnkshitJain** [comment]
It can be related to the problem of locales in Docker environment. @prasadtalasila Please have a look at this, and we can then verify it. 
https://blog.mkowalski.net/2016/05/17/solving-locale-issues-with-docker-containers/

**Comment by prasadtalasila** [comment]
@AnkshitJain Good find.
The problem seems real. The Ubuntu 16.04 OS has en_IN locale enabled, where as the execution node containers have the POSIX locale enabled. This explains the weird behavior in the containers.

I have run the locales commands on the host Ubuntu 16.04 operating system that students typically use. The results are as follows.
```shell
#show the enabled locale on the base machine
$ locale
LANG=en_IN
LANGUAGE=en_IN:en
LC_CTYPE="en_IN"
LC_NUMERIC="en_IN"
LC_TIME="en_IN"
LC_COLLATE="en_IN"
LC_MONETARY="en_IN"
LC_MESSAGES="en_IN"
LC_PAPER="en_IN"
LC_NAME="en_IN"
LC_ADDRESS="en_IN"
LC_TELEPHONE="en_IN"
LC_MEASUREMENT="en_IN"
LC_IDENTIFICATION="en_IN"
LC_ALL=

#next commands shows all the installed locales
$ locale -a
C
C.UTF-8
en_AG
en_AG.utf8
en_AU.utf8
en_BW.utf8
en_CA.utf8
en_DK.utf8
en_GB.utf8
en_HK.utf8
en_IE.utf8
en_IN
en_IN.utf8
en_NG
en_NG.utf8
en_NZ.utf8
en_PH.utf8
en_SG.utf8
en_US.utf8
en_ZA.utf8
en_ZM
en_ZM.utf8
en_ZW.utf8
POSIX
```

However, on the execution node containers, the locales are much limited. The results are as follows.
```shell
##show the enabled locale on execution node containers
# locale
LANG=
LANGUAGE=
LC_CTYPE="POSIX"
LC_NUMERIC="POSIX"
LC_TIME="POSIX"
LC_COLLATE="POSIX"
LC_MONETARY="POSIX"
LC_MESSAGES="POSIX"
LC_PAPER="POSIX"
LC_NAME="POSIX"
LC_ADDRESS="POSIX"
LC_TELEPHONE="POSIX"
LC_MEASUREMENT="POSIX"
LC_IDENTIFICATION="POSIX"
LC_ALL=

##show the installed locales
#locale -a
C
C.UTF-8
POSIX
```

**Summary** We need to install and activate en_IN locale in the execution node containers.  The following code, if put in Dockerfile of execution_nodes must get the "en_IN" locale into the docker containers as well.

```shell
RUN apt-get update
# make sure that locales package is available
RUN apt-get install --reinstall -y locales
# uncomment chosen locale to enable it's generation
RUN sed -i 's/# en_IN.UTF-8 UTF-8/en_IN.UTF-8 UTF-8/' /etc/locale.gen
# generate chosen locale
RUN locale-gen en_IN.UTF-8
# set system-wide locale settings
ENV LANG en_IN.UTF-8
ENV LANGUAGE en_IN  
ENV LC_ALL en_IN.UTF-8  
# verify modified configuration
RUN dpkg-reconfigure --frontend noninteractive locales
```  

I have run the configuration in one of the trial execution_nodes containers. The locales is set appropriately.

**Comment by prasadtalasila** [comment]
A sample error for mismatched locale is shown below.
```shell
$ javac *.java
iPhone6S.java:33: error: unmappable character for encoding UTF8
        // Gain access to the other student�s fields.
                                           ^
MotoG4Plus.java:33: error: unmappable character for encoding UTF8
        // Gain access to the other student�s fields.
                                           ^
2 errors
```

I get this error even after installing en_IN locale. There are umpteen number of locales; It seems Windows 10 has 220+ locales preinstalled. Thus BlueJ on Windows happily supports all these locales, but it becomes tricky for us to support all of these inside the execution nodes.

The saving grace is the helpful log message. Any problem with character encoding is clearly specified by the compiler in the log message.

One important observation. The locale settings made to the container after bash login are lost upon exit. When we login to the container, the locale is reset to the default POSIX locale.

In summary, it would be good to have en_IN locale as the default locale in the execution containers.

**Comment by AnkshitJain** [comment]
The locale settings will have to be initialized when the docker is built. The reason why the local changes made to a container is lost upon exit is given [here](https://askubuntu.com/questions/581458/how-to-configure-locales-to-unicode-in-a-docker-ubuntu-14-04-container) I'll include the changes in the Dockerfiles for the containers along with #199

**Comment by AnkshitJain** [comment]
This issue seems to with the Java compiler at the moment. Even with the POSIX locales, the evaluation works. The change needed was using `javac -encoding iso-8859-1 filename.java` instead of `javac filename` command for compilation. Java compiler assumes the code is in UTF8 format unless specified otherwise. Specifying a special encoding, solves the problem and the evaluation happens as expected.

**Comment by prasadtalasila** [comment]
@AnkshitJain The suggested solution solves the problem for java files only. We need to solve the problem for all programming languages. Hence the need to solve it with LOCALE initialization at the time of building the containers.

**Comment by prasadtalasila** [comment]
@AnkshitJain Please keep the utf-8 encoding format for characters. For testing purposes, you can modify the given test files to have one or two valid utf-8 characters in the javadoc comments. Such a test should pass.

**Comment by prasadtalasila** [comment]
PR #199 changes the locale to utf8, but we need a good test in tests/deployment_tests/execution-node.bats.

**Comment by prasadtalasila** [comment]
Make the deployment_tests/ directory structure comply with the new test structure.
